Add OPA Gatekeeper comparison page (#205)
kubewarden/docs: docs/comparisons/opa.md

| Policy enforcement modes                              | deny, warn                  | deny, warn                       |
| Deployment mode [[6]](#deployment-mode)               | single evaluation server    | multiple evaluation servers      |
| Background checks [[7]](#background-checks)           | ✅                          | ✅                               |

## Types of policies

Both OPA Gatekeeper and Kubernetes can write validation and mutation policies.

These policies can target any kind of Kubernetes resource, including Custom Resources.

## Writing policies

OPA Gatekeeper policies are written using [Rego](https://www.openpolicyagent.org/docs/latest/#rego).
Rego is a query language created by the Open Policy Agent project.

:::info
Rego can be used only to write validating policies. The policies performing mutation do not
use Rego, but instead use some ad-hoc rules defined via YAML (see [here](https://open-policy-agent.github.io/gatekeeper/website/docs/mutation)).
:::

Kubewarden allows policies to be written using different styles. Policy authors
can use both traditional programming languages (like Go, Rust and others) or Rego.

> Kubewarden's validating and mutating policies are written in the same way.

Reviewer: Rego is a language though, right. A domain specific language (¿I think?), but still ...... How about:

'can use both traditional programming languages (like Go, Rust and others) or the Rego domain specific language.'

I think it makes the difference more explicit. Maybe, hmmmm,

> Kubewarden's validating and mutating policies are written in the same way.

Author: I've updated the suggestion to reflect that, what do you think?

Despite using the same language, policies written for kube-mgmt are
not compatible with OPA Gatekeeper and vice versa.

Kubewarden can use Rego policies that have been written for both Open Policy Agent and
for OPA Gatekeeper. More information can be found [here](https://docs.kubewarden.io/writing-policies/rego/intro-rego).
:::

## Context aware

Sometimes a policy might need data about the current state of the cluster in order to take a
validation/mutation decision. For example, a policy validating Ingress resources might
need to know what are the other Ingress resources already defined inside of the cluster
to ensure no clash might happen.
This kind of policies are called "context aware".

Both OPA Gatekeeper and Kubewarden support this type of policies.

:::caution
When deploying OPA Gatekeeper, a Kubernetes administrator decides which type of
cluster data should be made available to the policies at evaluation time.

It's important to highlight how this data is then accessible by all the policies
deployed.

For example, if an OPA Gatekeeper policy requires access to Kubernetes Secrets,
all the other policies being deployed will be able to read this data as well.

> :::

Reviewer 2: I'm fine with this change.

On the other hand, Kubewarden provides a granular access to cluster resources.
Each policy has access only to the resources that the Kubernetes administrator
specified. Attempting to read unauthorized data is immediately blocked and

> reported to the cluster administrators.

Reviewer 2: I would say to remove the "caution" callout. For me, callouts try to catch my attention to something in the text. around. But here it is the actual comparision with Kubewarden. Therefore, I would remove it.

Author: I proposed the change

defined inside of the OPA Gatekeeper Custom Resource. This can be done using some 3rd
party tools.

Kubewarden policies have CI/CD pipelines like the ones of traditional microservices.
Usually their source code lives inside of a Git repository and then, using
traditional CI/CD systems, unit tests are ran against it. The unit tests are
written using the testing frameworks of the language used to write the policy.
Once all the tests are passing the policy is compiled to WebAssembly and pushed
to a container registry.
This kind of pipeline is usually maintained by the policy author.

Kubernetes administrators typically maintain other automation pipelines that react to
new releases of the policy (leveraging automation tools like
[Dependabot](https://docs.github.com/en/code-security/dependabot/working-with-dependabot),
[Renovate bot](https://www.mend.io/renovate/),
[updatecli](https://www.updatecli.io/) and others) or to changes to the
policy configuration.

The pipeline tests the policy against different types of requests. The testing can be done using
the [kwctl](https://github.com/kubewarden/kwctl) cli tool, without requiring a running

> Kubernetes cluster.

Reviewer 2: Is it not worth mentioning that the `kwctl` uses the same runtime used in production? I believe this will explicit that `kwctl` is a good tool to test the policies as close to production as possible.

Reviewer 3: +1 here.

Author: done

## Policy enforcement modes

> Both OPA Gatekeeper and Kubewarden can deploy policies using two different operation mode:

Reviewer: ```suggestion
Both OPA Gatekeeper and Kubewarden can deploy policies using two different operation modes:
```

> - `deny`: violation of a policy causes the request to be rejected
> - `warn`: violation of a policy does not cause the request to be rejected. The
>   violation is just logged for auditing purposes

Reviewer: ```suggestion
- `deny`: violation of a policy causes the request to be rejected
- `warn`: violation of a policy does not cause rejection. The
  violation is logged for auditing purposes
```

## Deployment mode

All the OPA Gatekeeper policies are evaluated by the same server.
On the other hand, Kubewarden allows multiple evaluation servers to be defined.
These servers are defined by a Custom Resource called `PolicyServer`.

> When declaring a Kubewarden policy, the Kubernetes administrator can then decide
> which `PolicyServer` is going to host it.

Reviewer: ```suggestion
When declaring a Kubewarden policy, the Kubernetes administrator decides
which `PolicyServer` will host it.
```

:::note

> The `PolicyServer` object is just a high level abstraction introduced by Kubewarden.
> Behind the scenes a Deployment with a specific replica size is created.

Reviewer: ```suggestion
The `PolicyServer` object is a high level abstraction introduced by Kubewarden.
Behind the scenes a `Deployment` with a specific replica size is created.
```

I'm not quite sure about whether to quote or decapitalize 'deployment'. Quotes probably good.

> Each `PolicyServer` can have a different replica size from the other ones.

Reviewer: ```suggestion
Each `PolicyServer` can have a different replica size from others.
```

:::

This allows interesting scenarios like the following ones:

- Deploy critical policies to a dedicated Policy Server pool
- Deploy the policies of a noisy tenant to a dedicated Policy Server pool

## Background checks

> As policies are added/removed/reconfigured the resources that are already inside
> of the cluster might suddenly become non-compliant.

Reviewer: ```suggestion
As policies are added, removed, and reconfigured the resources already inside
of the cluster might become non-compliant.
```

> Both OPA Gatekeeper and Kubewarden have a scanner that operates in the background.
> This scanner evaluates the resources already defined inside
> of the cluster and flags the one that are non-compliant.

Reviewer: ```suggestion
Both OPA Gatekeeper and Kubewarden have a scanner that operates in the background.
This scanner evaluates resources already defined inside
the cluster and flags non-compliant ones.
```
